### DISCUSSION TITLE
Gemeentelijke Architectuurvraagstukken

### DISCUSSION THREAD

**Discussion by marcvanandel** [kadaster-labs/sensrnet-home#245] [Pilot Gemeente]
Ten behoeve van implementatie en opschaling zijn er een aantal vraagstukken, architectuurvraagstukken die per gemeente of in het algemeen beantwoord moeten worden. Of in het algemeen besproken moeten worden, waar per gemeente keuzes over gemaakt moeten worden. Bijvoorbeeld:

- Welke scenario’s zijn er mogelijk qua hosting?
- Welke impact en consequenties zijn daar aan te verbinden?
- Waar lopen kleinere gemeenten tegenaan?
- Hebben die het moeilijker?
- Hoe moet de verbinding naar een andere organisatie gelegd worden?
- Financiële consequenties ook in beeld?
- Hoe gaan leveranciers in opdracht van een gemeente (of zelfs voor meerdere gemeenten) zelfstandig sensoren registreren in SensRNet?
  - Welke Registry Node?
  - Welke Organization?
  - Met welke rol?

In deze discussie is comment toe te voegen en tracht ik (@marcvanandel en actiehouder van dit punt) deze overwegingen bij te houden en te beantwoorden. Vul gerust aan en voeg vragen / comments toe!

**Comment by marcvanandel** [comment]
**Scenario's van hosting | Netwerk breed**
In de architectuur van het geheel heb ik al een suggestie geschreven over hoe opschaling zou kunnen plaatsvinden. Dit is beschreven in [network](https://kadaster-labs.github.io/sensrnet-home/Network/).

1. Het systeem is zo opgezet dat elke gemeente een eigen Registry Node kan hosten voor eigen gebruik, integratie met eigen systemen en door kan ontwikkelen voor aansluiting met hun eigen partners, hetzij bedrijven, burgers of andere overheden. Daarom een [open source project](https://kadaster-labs.github.io/sensrnet-home/KeyConcepts/#open-collaboration) en daarom het principe van [decentralisatie als gegeven](https://kadaster-labs.github.io/sensrnet-home/KeyConcepts/#decentralization). 
2. Toch is het mogelijk om met veel minder nodes het netwerk neer te zetten. In dat geval zullen partijen samen moeten werken op een beperkt aantal Registry Nodes. Dat zouden bijvoorbeeld alle provinciën kunnen doen voor alle gemeenten in hun jurisdictie. Of ... enige variatie / combinatie van aantallen en samenwerkingsverbanden.

**Scenario's van hosting | Registry Node**

Dit gaat over het überhaupt hosten van een Registry Node. Maar als je als gemeente (of bronhouder) besluit om een Registry Node te hosten, welke opties zijn er dan?

Uitgangspunt voor het hosten van een Registry Node is kubernetes (dit moet nog gedocumenteerd worden in een ADR; zie #183 ). Dat sluit aan bij de ontwikkelingen die er rondom cloud gaande zijn en ook bij VNG/Common Ground; zie [Haven](https://haven.commonground.nl/). Vanuit dat perspectief wordt er vanuit gegaan dat elke gemeente (straks / ooit) een Haven-compliant kubernetes cluster / cloud omgeving tot haar beschikking heeft ... en dan is installatie en hosting van een Registry Node geen issue (meer). Dat valt onder 'gewoon werk'.

MAAR ... Haven is dan wel Common Ground standaard ... maar nog geen 'gemeente standaard', oftewel, veel gemeenten lopen er tegenaan dat zij (nog) geen Haven-compliant cloud platform tot hun beschikking hebben. Tja, wat dan?

De SensRNet componenten zijn containers, op zichzelf staande 'doosjes' waarvan de installatiescripts uitgaan van Kubernetes (volgens Haven-standaard). Maar deze containers zijn daar niet van afhankelijk. Deze kunnen ook als containers gestart worden in een omgeving zoals [Docker](https://www.docker.com/). Lokale software ontwikkeling vindt ook plaats mbv `docker-compose`, welke ook met niet al te grote moeite te gebruiken kunnen zijn voor een pilot omgeving obv Docker. Uiteindelijk zijn containers 'ook maar boxes/dozen' om processen heen. De instructie om deze processen te starten, staan in de container beschrijvingen en kunnen dan ook zonder de container als black box er omheen, los gestart worden. Uitgangspunt is een Linux omgeving ... maar zelfs op Windows zou dat gestart moeten kunnen worden.

Aangezien het uitgangspunt Kubernetes volgens Haven-standaard is gekozen, is er geen / weinig ondersteuning beschikbaar voor andere omgevingen. Met inbreng van eigen kennis is vrijwel alles mogelijk ... maar de kosten zijn dan vooral afhankelijk van eigen inbreng en expertise.

**Scenario's van hosting | Join someone else's Registry Node**

Dit scenario is een scenario waarin je als bronhouder niet zelf een Registry Node host, maar 'joined', meedoet in de Registry Node van een ander. In het netwerk stukje hierboven wordt deze optie al benoemd ... en eigenlijk hoort dit niet meer bij hosting, want je host dan zelf niets meer. Nou ja ... de koppeling naar een eigen 'user identity system' blijft wel mogelijk. De componenten van de Registry Node ondersteunen meerdere 'Identity Providers' zodat gebruikers SSO (Single Sign On) kunnen registreren (automatisch) en zo in een Registry Node aan de slag kunnen. Dit scenario gaat dan uit van hosting door een andere partij en (dus) wel een eigen user identity systeem.

Zie ook #230

**Reply by marcvanandel** [reply]
'Join' kan technisch wel, maar is dit wel geaccepteerd door de cloud leveranciers? Dit lijkt nog geen sinecure ... en dat is begrijpelijk.

In het geval dat een gemeente een contract heeft met een public cloud vendor - vaak Microsoft Azure - en vervolgens een tweede contract met een andere vendor om daar bovenop een Haven-compliant / Common Ground services laag aan te bieden, dan is multi-tenancy al een ingewikkeld concept. Dat wordt nóg complexer als vervolgens de gemeente kiest om een Registry Node te hosten en _binnen_ die Registry Node andere gemeenten toegang te gaan geven. Dat wordt een multi-multi-tenancy en wie heeft dan een SLA met wie en wie betaalt wie (en is dat fair)? Oef! Deze constructies zijn toch wel ingewikkeld 😏

**Reply by marcvanandel** [reply]
Gemeenten worstelen nog veel met het beschikbaar hebben van een Haven-compliant platform. Doen zij dat zelf? Halen zij dat uit de markt? In welke contractuele vorm dan? 'Common Ground as a Service' bovenop public cloud leveranciers? Dit lijkt een veel gebruikte constructie ...

Ik zal input bij het Haven-team / VNG-Realisatie 'Core Team' (NLX+Haven) gaan ophalen hierover...

**Reply by marcvanandel** [reply]
Er zijn nu 8 referentie implementaties van Haven beschikbaar (waar? tip uit Common Ground Slack discussie) Tegelijk wordt dit punt wel herkent: Een Haven-compliant cloud platform is niet zomaar geregeld door een gemeente, met name door contracten, cloud transitie problematieken, ontbreken van kennis en/of regie (en zelden de techniek zelf)

**Reply by marcvanandel** [reply]
Het concept van 'join'n klinkt als 'multi tenancy _binnen_ een applicatie' ... wat een ouderwets model is. Multi-tenancy behoort ingebakken te zijn in de SaaS dienst die je afneemt. Maar ligt nu juist de crux van dit vraagstuk: Wil je als gemeente een SaaS dienst afnemen voor je sensorenbeheer? Of wil je maximaal kunnen integreren in je bestaande (cloud) landschap? De architectuur van SensRNet is met name gericht / bedoeld voor de tweede optie, omdat de eerste optie daarmee ook mogelijk is. Dat wil zeggen, op dit moment (MVP plus fase) is er geen beoogde partij / (cloud) locatie voor het hosten van een SaaS oplossing. Dat moet nog uitgewerkt cq uitgedacht worden. Dit is mede afhankelijk van de vraag van gemeenten.

ERGO: Beste gemeente, wilt u een applicatie integreren in uw eigen applicatie landschap of wilt u een SaaS dienst afnemen?

-> Applicatie integreren in eigen applicatie landschap => Haven omgeving + huidige applicatie (scripts)
-> SaaS dienst afnemen ... => Wie host dat? Wie zou dat moeten hosten? Hoeveel mag dat kosten?

**Reply by kad-busses** [reply]
Link naar referentie implementaties: https://haven.commonground.nl/docs/aan-de-slag

**Reply by mikegroels** [reply]
Om nog antwoord te geven op je vraag Marc:
[Tilburg] De voorkeur gaat uit naar een SAAS applicatie die we met API’s kunnen koppelen\integreren in ons eigen applicatie landschap.
Argumenten:
•	We hanteren een SaaS first strategie
•	SensRNet is heeft [weinig] additionele functionaliteit.
•	Op basis van het vorige punt verwacht ik weinig geavanceerde integraties [als in: koppelingen]. Op dit moment zie ik alleen het vullen en uitlezen van de database.

**Reply by BramWithaar** [reply]
Vanuit Nijmeegs perspectief sluit ik me aan bij wat Mike aangeeft. 

De veronderstelling dat eigen applicatielandschap bij een gemeente gelijk staat aan Kubernetes/Haven klopt niet.
Ik zou verwachten dat integratie met andere systemen op basis van api's gebeurd; het opnemen van deze containers in een eigen omgeving voegt dan weinig toe.
Het zelf beheren van een oplossing als deze door een gemeente voegt ook weinig tot niks toe aan de waarde van de oplossing, of aan dienstverlening voor een inwoner. Dan ligt het al snel voor de hand om het als SaaS af te willen nemen.

In de Slack heb ik eerder dit, op persoonlijke titel, geplaatst:
de vraag is wat mij betreft waarom je een oplossing als dit niet centraal aanbied (SaaS). Gemeenten nemen steeds meer af als SaaS, en lokale ict landschap wordt daarmee kleiner. Een lokaal kubernetes cluster gaat tegen die stroom in (en vraagt kennis die in de meeste gemeenten niet aanwezig is).
Dan komt een gemeente bij de volgende optie uit; een kubernetes cluster bij een leverancier. Maar waarom zouden we voor dit soort applicaties 300+ kubernetes clusters moeten hebben, met ook 300x inkoop gedoe, en als een van die apps dingen met DigiD doet ook 300x een audit (en echt heel veel gedoe, en kosten, met de hosting partij daarover om dat geregeld te krijgen).
Daarnaast is het, door zo te richten op containers, lastiger om gebruik te maken van meer abstracte diensten (managed services) van cloudproviders (zie bv ook de nieuwe container omgeving op Azure https://samenorganiseren.slack.com/archives/CGCPJE599/p1635878185000700 , maar ook loadbalancers, queue-services, logging, storage, functions-as-service, encryptie, search, api-gateway).
Mijn inschatting is dat als je veel gemeenten wil bereiken met een app, en het voor het collectief van gemeenten ook nog een beetje kosten efficiënt wil doen, en wil aansluiten bij de beweging die gemeenten maken, je je app zoveel mogelijk als SaaS moet aanbieden. En houd dan ook niet vast aan Kubernetes als onderlaag, gebruik containers waar nodig, maar maak ook gebruik van wat een provider aan managed services heeft te bieden.
Dit is heel goed te combineren met het basis idee van de Cloud Based Infra (CBI).

Als laag-hangend fruit, Het 'simpel' draaien als containers op een ander platform dan Kubernetes zou al veel simpeler zijn als er iets als dit https://docs.nlx.io/reference-information/environment-variables aan documentatie beschikbaar is. Dat zou het voor ons potentieel (afhankelijk van aantal containers en hoe opslag geregeld kan worden) al wat makkelijker maken, omdat we wel een container platform hebben, maar geen Kubernetes en/of Helm gebruiken.

**Comment by marcvanandel** [comment]
**Registry Node vs Organization**

Een Registry Node gaat over het hosten van software om metadata rondom sensoren te kunnen registeren / vast te leggen. Eén van de onderdelen van de metadata is de `Organization` die de eigenaar (`Owner` of andere rol) is van de sensor (eigenlijk: `Device`). Een Registry Node is NIET 1-op-1 gelijk met een Organization. Dat zijn losstaande concepten.

Bijvoorbeeld kan een gemeente een Registry Node hosten en daar zichzelf registreren als Organization. Daarnaast kunnen alle bedrijven (en later wellicht alle burgers) zichzelf (*) óók registreren als Organization. Zo kunnen er sensoren (devices) van diverse en vele partijen worden geregistreerd met één Registry Node.

_* NOTE Hiervoor moet wel in de Registry Node ondersteund worden dat gebruikers van die partijen zich kunnen registreren op een of andere manier. Hiervoor zijn nog geen specificaties uitgewerkt en dat is (dus) ook nog niet beschikbaar als standaard functionaliteit._

Een ander scenario is - zoals Gemeente Amsterdam aan het ontwikkelen is voor de Meldplicht - dat een gemeente andere partijen gaat registreren en hun sensoren aan die Organizations gaat toevoegen. Dan treedt de gemeente op als 'penhouder' cq bronhouder van de metadata, terwijl de sensoren als eigenaar gekoppeld en gepubliceerd worden onder de partijen die daadwerkelijk eigenaar zijn van die sensoren.

_NOTE Ook dit scenario is nog te onduidelijk qua specificaties om daadwerkelijk te realiseren in de software en (dus) is (ook) dit geen standaard functionaliteit die nu al beschikbaar is._

Omdat Registry Node en Organization losstaande concepten zijn, kunnen er ook bijzondere scenario's ontstaan, zoals:

- Een gemeente begint als Organization in een Registry Node van een andere bronhouder en besluit om toch zelf een Registry Node te gaan hosten (incl. integraties en data migraties)
- Een gemeente begint als Organization in een eigen Registry Node en om redenen besluit zij over te stappen op de Registry Node van een andere bronhouder
- Door de fusie van twee (of meer) gemeenten worden Registry Nodes samengevoegd en wellicht de Organizations ook

Voor deze scenario's is extra functionaliteit nodig om migraties en 'overdracht van eigendom van de data' mogelijk te maken. Zie (oa) #238

**Reply by marcvanandel** [reply]
Een toekomstbeeld dat wat verder weg ligt, zou meer op de blockchain manier van werken gebaseerd kunnen zijn...

Stel dat elke bronhouder een eigen 'wallet' heeft voor zijn eigen Organization data en devices data, zoals elke gebruiker een eigen wallet heeft voor z'n eigen BitCoins. Dan zouden er diverse 'apps' kunnen zijn en de Registry Node is daar dan één van, een server-side 'app' met uitgebreide mogelijkheden voor integraties. Dan zou daarnaast een lokaal installeerbare app ontwikkeld kunnen worden en een mobiele app en ... door de markt ontwikkelde apps waarmee gebruikers hun wallet op de blockchain kunnen beheren.

Dit idee sluit aan bij een verdere ontwikkeling richting blockchain en een breder gebruik van blockchain - zie ook #264. Op dit moment wordt blockchain (of eigenlijk distributed ledger technology - DLT) 'slechts' gebruikt als technologische laag voor het sensorenregister, maar dat zou in de toekomst mogelijk uitgebreid kunnen worden ...

**Comment by marcvanandel** [comment]
Hoe wordt de rolverdeling tussen de gemeente en een landelijke voorziening?

**Reply by marcvanandel** [reply]
De architectuur onderkent nodes in een netwerk. Uiteraard is de Registry Node daarvan een essentiële, want daarmee wordt de metadata van sensoren beheerd. Zonder deze nodes is er geen data om te delen in het netwerk. Naast de Registry Nodes kunnen er ook ander soortige nodes deelnemen aan het netwerk, bijvoorbeeld een Publishing Node. Door één van die Publishing Nodes te benoemen tot centrale / landelijke voorziening, is een centraal / landelijk punt gecreëerd waar 'men' (in het algemeen) terecht kan voor informatie over sensoren. Dit hoeft echter niet de enige publicatie plaats en vorm te zijn. Er kunnen ook andere / meer publishing nodes van verschillende vormen toegevoegd worden aan het netwerk. De architectuur beperkt dat niet.

Hoe wordt de rolverdeling tussen de gemeente en een landelijke voorziening? Een landelijke voorziening in de zin van een publicatie van sensoren is 'gewoon een node in het netwerk' ... en zo is dan ook die rolverdeling. Dit kan (moet!) in het consortium van enige vorm afgesproken en vastgelegd worden.

Hoe wordt de rolverdeling tussen de gemeente en een landelijke voorziening in de zin van de community / samenwerkingsverband van alle partijen? Tja, daar zijn vele vormen mogelijk! Er is hier geen voorgeschreven vorm / verdeling / afspraak voor. Dit moet afgestemd, besproken, uitgewerkt worden in hoe de betrokken partijen dit graag zouden willen zien. De architectuur maakt alles (nou ja, veel 😉) mogelijk!

**Comment by marcvanandel** [comment]
**Asset management systeem vs SensRNet Registry Node**

In alle gemeenten (en bij bronhouders in het algemeen) zijn sensoren ook assets. Deze moeten worden beheerd en daarvoor is vaak een aparte afdeling (of meerdere afdelingen, afhankelijk van het soort sensor en omvang van de bronhouder) verantwoordelijk. En daarvoor wordt vaak een asset management systeem gebruikt. Tegelijk is de metadata in SensRNet daar een afspiegeling van. Of is het asset management systeem een afspiegeling van de metadata in SensRNet?

Met andere woorden:
- **Is het asset management system leidend of is de SensRNet Registry Node leidend?**
- **Hoe is de integratie / mapping op IDs / referenties tussen het asset management systeem en de SensRNet Registry Node?**

(oa bij gemeente Utrecht zijn dit actuele vragen; bij welke gemeenten nog meer?)

**Reply by mikegroels** [reply]
Ja, die vragen leven bij ons in Tilburg ook.
Op dit moment overlappen ze half en zal nog moeten worden afgewogen (afhankelijk van de beschikbaarheid en werking van eventuele koppelingen) welk systeem leidend is. SensRNet zou niet het complete asset management hoeven af te dekken (met contracten, afspraken met leveranciers en de verbinding naar eventuele andere assets).
